src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/BuildUpToDateCheck.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.VisualStudio.ProjectSystem.Build;
+using Microsoft.VisualStudio.ProjectSystem.Logging;
+using Microsoft.VisualStudio.ProjectSystem.References;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem
+{
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    [Export(typeof(IBuildUpToDateCheckProvider))]
+    internal class BuildUpToDateCheck : IBuildUpToDateCheckProvider
```

Comment:
Also, I would consider either splitting up into multiple `IBuildUpToDateCheckProvider` that cover smaller checks. This implementation is trying to be a catch-all check and may become difficult to maintain in the future.

If you don't want multiple implementations, then I would at least try to containerize each individual check performed here into helper classes, or at least individual functions.